storage_service/ownership: handle requests when tablets are enabled

pwrobelse opened the pull request:
Before this change, when user tried to utilize
'storage_service/ownership/{keyspace}' API with
keyspace parameter that uses tablets, then internal
error was thrown. The code was calling a function,
that is intended for vnodes: get_vnode_effective_replication_map().

This commit introduces graceful handling of such scenario and
extends the API to allow passing 'cf' parameter that denotes
table name.

Now, when keyspace uses tablets and cf parameter is not passed
a descriptive error message is returned via BAD_REQUEST.
Users cannot query ownership for keyspace that uses tablets,
but they can query ownership for a table in a given keyspace that uses tablets.

Also, new tests have been added to test/rest_api/test_storage_service.py and
to test/topology_experimental_raft/test_tablets.py in order to verify the behavior
with and without tablets enabled.

Fixes: https://github.com/scylladb/scylladb/issues/17342

pwrobelse commented on test/rest_api/test_storage_service.py:
```diff
@@ -586,3 +594,20 @@ def test_range_to_endpoint_map_with_table_param(cql,  rest_api, tablets_enabled)
+        expected_error_reason = f"storage_service/ownership is not supported for keyspace {keyspace} because it uses tablets."
+        assert expected_error_reason == actual_error_reason
+
+@pytest.mark.xfail(reason="rest_api suite doesn't support tablets yet (#17338), run test manually")
+def test_ownership_tablets_enabled_keyspace_param_missing(cql, this_dc, rest_api):
+    with new_test_keyspace(cql, f"WITH REPLICATION = {{ 'class' : 'NetworkTopologyStrategy', '{this_dc}' : 1 }} AND TABLETS = {{ 'enabled': true }}") as keyspace:
+        resp = rest_api.send("GET", f"storage_service/ownership")
+        resp.raise_for_status()
```
I am not entirely sure about this test case. It passes, but there is a keyspace that uses tablets.

Should such request be also discarded like the one above?

avikivity replied:
What does it return for the vnodes case?

pwrobelse replied:
Array of pairs: (node: IP, ownership: float) is returned.

Currently, I do not see any difference in results when keyspace is not specified when using this API. 

This is a sample result for **_storage_service/ownership_**:
`[{'key': '127.7.51.50', 'value': '1.0000001'}]`

avikivity replied:
Try a keyspace with RF=1 and initial_tablets=1 (when using tablets)

With vnodes, you should get node1->0.5 node2->0.5 (won't be exactly 0.5)
With tablets, you should get nodeX->0 nodeY->1 (can't predict if X is 1 or 2)

avikivity replied:
btw reaching ownership > 1 is impressive

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6790/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6790/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8013/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6790/)
#### Build Details:
- Duration: 1 hr 42 min
- Builder: spider8.cloudius-systems.com

avikivity commented:
For range_to_endpoint_map (3842bf18a7cfb09d43209702ebed6414f098ea0e), someone extended it with a table parameter. We could do the same here and ask ownership about a particular table. That's useful to see how a table is distributed among nodes (esp. a smaller table that hasn't reached all nodes yet or suffers from quantization).

pwrobelse commented:
> For range_to_endpoint_map ([3842bf1](https://github.com/scylladb/scylladb/commit/3842bf18a7cfb09d43209702ebed6414f098ea0e)), someone extended it with a table parameter. We could do the same here and ask ownership about a particular table. That's useful to see how a table is distributed among nodes (esp. a smaller table that hasn't reached all nodes yet or suffers from quantization).

Okay, sure. I will try to do that. To proceed, I need to ensure that `effective_replication_map` supports similar functionality that `vnode_effective_replication_map::get_ranges(inet_address ep)` provides.  The mentioned member function is specific for the derived class for vnodes and it is used in `storage_service::effective_ownership()`.

pwrobelse commented:
> For range_to_endpoint_map ([3842bf1](https://github.com/scylladb/scylladb/commit/3842bf18a7cfb09d43209702ebed6414f098ea0e)), someone extended it with a table parameter. We could do the same here and ask ownership about a particular table. That's useful to see how a table is distributed among nodes (esp. a smaller table that hasn't reached all nodes yet or suffers from quantization).

I was working on the required changes. It appears, that the code, which I prepared allows to do that. However, before pushing it I wanted to double-check if its design is valid.

1. `storage_service::effective_ownership()` that calculates ownership in storage service utilizes a function, that is specific for `vnode_effective_replication_map` - `dht::token_range_vector get_ranges(inet_address ep)`.
2. In my change I added the requirement related to implementation of this function to `effective_replication_map` interface - it became pure-virtual function.
3. Both `vnode_effective_replication_map` and `tablet_effective_replication_map` provide implementations of this function.
4. Because of (3) now `storage_service::effective_ownership()` uses `effective_replication_map` - depending on the parameters passed by the user of this function, it assigns `ks.get_vnode_effective_replication_map()` or `cf.get_effective_replication_map()`.

I adjusted the python tests to verify the returned ownership - because the test uses single node and RF=1 ownership equals 1 is returned. All tests are passing right now.

However, the implementation of `tablet_effective_replication_map::get_ranges(inet_address ep)` uses almost identical logic to `vnode_effective_replication_map::get_ranges(inet_address ep)`. Firstly, it gets `tm.sorted_tokens()`, then it iterates over them and for each token it queries `get_natural_endpoints(tok)` and scans them - if the required endpoint is found, then it inserts token range to the result container.

So, my questions are as follows:
1. Can `get_ranges(endpoint)` become pure virtual function and part of `effective_replication_map`? Is it aligned with the remaining design decisions?
2. Is there a better way to get the ranges from `tablet_effective_replication_map` to avoid code duplication?
3. Should I add also tests, which use more than 1 node and different replication factor to get different results of ownership?
4. Is the approach, that I described valid? I mean, I can see that in case of tablets enabled and passed ks+cf the ownership is 1, but I decided to ask to ensure, that in general it is correct.

Apologies for the basic questions - I am still learning about the project and I am not sure what are the previous design decision regarding effective replication map :sweat:

avikivity commented:
Push the new code, so I can comment on it

pwrobelse commented:
> Push the new code, so I can comment on it

Done - the new patch has been just uploaded

pwrobelse commented on locator/tablets.cc:
Both functions added in this file look like a copy-paste. I don't like the way how I did it in the current patch.

avikivity commented on locator/tablets.cc:
```diff
@@ -523,6 +553,34 @@ class tablet_effective_replication_map : public effective_replication_map {
+        dht::token_range_vector ret;
+
+        auto prev_tok = sorted_tokens.back();
+        for (const auto& tok : sorted_tokens) {
```
It's wrong to use sorted_tokens here, they're only relevant for vnodes. Use get_tablet_map() to get the table-specific layout.

You can use first_tablet/next_tablet to iterate. Ideally you'd use for_each_tablet() but that returns a future, so add a FIXME to change get_ranges() to return a future.

avikivity replied:
btw, you won't need to use insert_token_range_to_sorted_container_while_unwrapping since tablets don't wrap around the ring (end of tablet always > start of tablet).

pwrobelse replied:
Sure, corrected the implementation as follows:

```
+    virtual dht::token_range_vector get_ranges(inet_address ep) const override {
+        dht::token_range_vector ret;
+
+        auto&& tablets = get_tablet_map();
+        tablets.for_each_tablet([&](locator::tablet_id id, const tablet_info& info) -> future<> {
+            auto endpoints = get_natural_endpoints(tablets.get_last_token(id));
+            auto should_add_range = std::find(std::begin(endpoints), std::end(endpoints), ep) != std::end(endpoints);
+
+            if (should_add_range) {
+                ret.push_back(tablets.get_token_range(id));
+            }
+
+            return make_ready_future<>();
+        }).get();
+
+        return ret;
+    }
```

To enable usage of the new implementation I had to adjust its caller - `storage_service::effective_ownership()`.

It uses sorted tokens to describe the ownership:
`const auto token_ownership = dht::token::describe_ownership(tm.sorted_tokens())`

I had to create `token_ownership` in a different way. Instead of using `sorted_tokens()` I had to get tokens vector from tablet map. Right now I am extending the tests to check if it works correctly in all cases - the basic case with single node and RF=1 passed.

avikivity commented on test/rest_api/test_storage_service.py:
```diff
@@ -586,3 +611,28 @@ def test_range_to_endpoint_map_with_table_param(cql,  rest_api, tablets_enabled)
 
                 assert endpoint == expected_endpoint, f"Unexpected endpoint={endpoint} for token_range={token_range}. Expected endpoint was {expected_endpoint}"
+
+@pytest.mark.xfail(reason="rest_api suite doesn't support tablets yet (#17338), run test manually")
```
It will just break if we add xfails like that. @bhalevy let's run the API suite with tablets.

bhalevy replied:
cc @tchaikov

denesb replied:
https://github.com/scylladb/scylladb/pull/17474

pwrobelse replied:
Fixed - introduced usage of the new fixtures.

avikivity commented on test/rest_api/test_storage_service.py:
```diff
@@ -586,3 +611,28 @@ def test_range_to_endpoint_map_with_table_param(cql,  rest_api, tablets_enabled)
+        with new_test_table(cql, keyspace, 'p int PRIMARY KEY') as table:
+            cf = table.split('.')[1]
+            resp = rest_api.send("GET", f"storage_service/ownership/{keyspace}", params={"cf": cf})
+            verify_ownership(resp=resp, expected_ip=rest_api.host, expected_ownership=1.0, delta=0.001)
```
Need two nodes and initial_tablets = 1 to get an imbalance (unlike vnodes which will always be balanced).

pwrobelse replied:
Introduced additional TCs to `./test/topology_experimental_raft/test_tablets.py` to setup more than one node and verify how the ownership behaves.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6819/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6819/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/4/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6819/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8043/)
#### Build Details:
- Duration: 2 hr 39 min
- Builder: spider3.cloudius-systems.com

avikivity commented:
btw title is now wrong

pwrobelse commented on service/storage_service.cc:
```diff
@@ -3101,7 +3107,32 @@ future<std::map<gms::inet_address, float>> storage_service::effective_ownership(
+            const replica::keyspace& ks = ss._db.local().find_keyspace(keyspace_name);
+            if (!ks.uses_tablets()) {
+                return dht::token::describe_ownership(tm.sorted_tokens());
+            } else {
```
I need to find a better place for this logic. Also, I will double-check if it is correct. Right now the tests pass.

Basically, I had to calculate `token_ownership` on basis of tokens from tablet map instead of sorted tokens.

pwrobelse commented:
> btw title is now wrong

Done - it has been adjusted. Thanks :+1:

pwrobelse commented on api/storage_service.cc:
```diff
@@ -1367,8 +1367,18 @@ void set_storage_service(http_context& ctx, routes& r, sharded<service::storage_
     ss::get_effective_ownership.set(r, [&ctx, &ss] (std::unique_ptr<http::request> req) {
-        auto keyspace_name = req->param["keyspace"] == "null" ? "" : validate_keyspace(ctx, req->param);
-        return ss.local().effective_ownership(keyspace_name).then([] (auto&& ownership) {
+        auto keyspace = req->param["keyspace"] == "null" ? "" : validate_keyspace(ctx, req->param);
```
Hi @avikivity, I have a question just to ensure that I understand this line of code correctly.

Is the following path some **special** kind of ownership query?
`/storage_service/ownership/null`

I wonder, because at glance I thought that such scenario would be executed when keyspace parameter is not specified - that was the first thing that came to my mind when I saw `null`.

However, it seems, that in case of `/storage_service/ownership/` a completely different path is taken.

According to api/api-doc/storage_service.json:
```
{
         "path":"/storage_service/ownership/",
         "operations":[
            {
               "method":"GET",
               "summary":"The mapping from token -> % of cluster owned by that token",
               "type":"array",
               "items":{
                  "type":"mapper"
               },
               "nickname":"get_ownership",
               "produces":[
                  "application/json"
               ],
               "parameters":[
               ]
            }
         ]
      }
```
and
```
"path":"/storage_service/ownership/{keyspace}",
         "operations":[
            {
               "method":"GET",
               "summary":"Effective ownership is % of the data each node owns given the keyspace",
               "type":"array",
               "items":{
                  "type":"mapper"
               },
               "nickname":"get_effective_ownership",
               "produces":[
                  "application/json"
               ],
               "parameters":[
                  {
                     "name":"keyspace",
                     "description":"The keyspace to fetch information about",
                     "required":true,
                     "allowMultiple":false,
                     "type":"string",
                     "paramType":"path"
                  },
                  {
                     "name":"cf",
                     "description":"Column family name",
                     "required":false,
                     "allowMultiple":false,
                     "type":"string",
                     "paramType":"query"
                  }
               ]
            }
         ]
```

So, as far as I understand in case of `/storage_service/ownership` ss::get_ownership is used, whether in case of `/storage_service/ownership/keyspace` ss::get_effective_ownership is invoked. 


Perhaps, if `/storage_service/ownership/null` is a special query, then additional tests need to be added to this PR.


Also, in spite of adding tests that execute `/storage_service/ownership` it seems, that they are not valid when tablets are enabled. `ss::get_ownership` is called and it invokes `storage_service::get_ownership()` that uses only `sorted_tokens()` - so, as far as I understand it does not support tablets. Do I understand it correctly?

Perhaps, this PR needs to be enhanced by taking into account also the other handler.

denesb replied:
> Is the following path some special kind of ownership query?
> /storage_service/ownership/null

Yes. This is used by `nodetool status`, when invoked without a keyspace argument.

denesb replied:
The difference between `ss::ownersip()` and `ss::effective_ownership()` is that the latter takes into account how data is replicated in datacenters. The former doesn't.

For tablets, we will need a completely different code path for both: this implementation would use `system.tablets` and not the sorted tokens, as the basis of calculating ownership. Further thinking about it, I think with tablets, we can only provide meaningful implementation for `ss::effective_ownership()`. `ss:ownership()` has even less meaning with tablets than it has with vnodes (and it is already pretty useless with vnodes).

It looks like we will need some changes to `nodetool status` to adjust for tablets. I'm in the process of re-implementing it in C++ currently (we are only doing tablet-related changes in the native nodetool, not the legacy one).

pwrobelse replied:
Hi @denesb - I introduced a new if statement in handler of `storage_service/ownership` that checks if any of keyspaces uses tablets - in such case BAD_REQUEST is returned. To verify the new logic an additional test case has been introduced.

Is it okay? Could you take a look at it?

bhalevy replied:
Would it make sense to push the functionality into a vitual method in effective_replication_map?

pwrobelse replied:
> Would it make sense to push the functionality into a vitual method in effective_replication_map?

Hi @bhalevy - could you specify which functionality? Are you referring to `storage_service/ownership/null`?

bhalevy replied:
> > Would it make sense to push the functionality into a vitual method in effective_replication_map?
> 
> Hi @bhalevy - could you specify which functionality? Are you referring to `storage_service/ownership/null`?

no, I meant the following functionality:

> For tablets, we will need a completely different code path for both: this implementation would use system.tablets and not the sorted tokens, as the basis of calculating ownership.

bhalevy replied:
Anyhow, making get_ranges virtual seem to go in exactly that direction.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6911/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6911/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8114/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/77/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6911/)
#### Failed Tests (9/23420):
- [topology_experimental_raft.test_tablets.debug.4](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6911/testReport/junit/%28root%29/non-boost%20tests/topology_experimental_raft_test_tablets_debug_4) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_experimental_raft.test_tablets.debug.4)
- [topology_experimental_raft.test_tablets.debug.5](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6911/testReport/junit/%28root%29/non-boost%20tests/topology_experimental_raft_test_tablets_debug_5) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_experimental_raft.test_tablets.debug.5)
- [topology_experimental_raft.test_tablets.debug.6](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6911/testReport/junit/%28root%29/non-boost%20tests/topology_experimental_raft_test_tablets_debug_6) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_experimental_raft.test_tablets.debug.6)
- [test_storage_service_api_even_ownership_keyspace_and_table_params_used](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6911/testReport/junit/%28root%29/test_tablets/test_storage_service_api_even_ownership_keyspace_and_table_params_used) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_storage_service_api_even_ownership_keyspace_and_table_params_used)
- [test_storage_service_api_uneven_ownership_keyspace_and_table_params_used](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6911/testReport/junit/%28root%29/test_tablets/test_storage_service_api_uneven_ownership_keyspace_and_table_params_used) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_storage_service_api_uneven_ownership_keyspace_and_table_params_used)
- [test_storage_service_api_even_ownership_keyspace_and_table_params_used](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6911/testReport/junit/%28root%29/test_tablets/test_storage_service_api_even_ownership_keyspace_and_table_params_used_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_storage_service_api_even_ownership_keyspace_and_table_params_used)
- [test_storage_service_api_uneven_ownership_keyspace_and_table_params_used](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6911/testReport/junit/%28root%29/test_tablets/test_storage_service_api_uneven_ownership_keyspace_and_table_params_used_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_storage_service_api_uneven_ownership_keyspace_and_table_params_used)
- [test_storage_service_api_even_ownership_keyspace_and_table_params_used](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6911/testReport/junit/%28root%29/test_tablets/test_storage_service_api_even_ownership_keyspace_and_table_params_used_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_storage_service_api_even_ownership_keyspace_and_table_params_used)
- [test_storage_service_api_uneven_ownership_keyspace_and_table_params_used](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6911/testReport/junit/%28root%29/test_tablets/test_storage_service_api_uneven_ownership_keyspace_and_table_params_used_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_storage_service_api_uneven_ownership_keyspace_and_table_params_used)
#### Build Details:
- Duration: 2 hr 6 min
- Builder: spider5.cloudius-systems.com

raphaelsc commented on locator/tablets.hh:
```diff
@@ -348,6 +348,9 @@ public:
 
+    /// Returns a vector of contiguous tokens where each pair describes a range of tablet.
+    seastar::future<std::vector<token>> get_flat_token_ranges() const;
+
```
seastar namespace is used in most of places, so no need to specify it.

pwrobelse replied:
Thanks, fixed!

raphaelsc commented on locator/tablets.cc:
```diff
@@ -220,6 +223,23 @@ dht::token_range tablet_map::get_token_range(tablet_id id) const {
 
+seastar::future<std::vector<token>> tablet_map::get_flat_token_ranges() const {
+    auto tokens = make_lw_shared<std::vector<token>>();
+    return for_each_tablet([this, tokens](locator::tablet_id id, const locator::tablet_info& info) -> future<> {
```
isn't it better to simplify it with coroutines?

pwrobelse replied:
Fixed - thanks! New patch will be uploaded soon.

raphaelsc commented on locator/tablets.cc:
```diff
@@ -220,6 +223,23 @@ dht::token_range tablet_map::get_token_range(tablet_id id) const {
 
+seastar::future<std::vector<token>> tablet_map::get_flat_token_ranges() const {
+    auto tokens = make_lw_shared<std::vector<token>>();
+    return for_each_tablet([this, tokens](locator::tablet_id id, const locator::tablet_info& info) -> future<> {
```
avoid reallocations with tokens.reserve(tablet_count())

raphaelsc replied:
tablet_count() + 1.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6961/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6961/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8144/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6961/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/107/)
#### Build Details:
- Duration: 2 hr 3 min
- Builder: spider5.cloudius-systems.com

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6967/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6967/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8148/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/111/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6967/)
#### Build Details:
- Duration: 2 hr 8 min
- Builder: spider1.cloudius-systems.com

denesb commented on locator/tablets.cc:
```diff
@@ -220,6 +223,25 @@ dht::token_range tablet_map::get_token_range(tablet_id id) const {
+    tokens.reserve(tablet_count() + 1u);
+
+    for (auto id : tablet_ids()) {
+        auto range = get_token_range(id);
```
Each tablet has exactly one token. Use `tablet_map::get_last_token()` to obtain this.

denesb replied:
Explanation: in both the ring and tablets, there are individual tokens, which form a ring. In the ring, the range of a given vnode/tablet extends from the last token of the previous vnode/tablet to that of the current one. This is implicit, so you don't need to special handle the first range.

denesb commented on locator/tablets.hh:
```diff
@@ -348,6 +348,9 @@ public:
     dht::token_range get_token_range(tablet_id id) const;
 
+    /// Returns a vector of contiguous tokens where each pair describes a range of tablet.
+    future<std::vector<token>> get_flat_token_ranges() const;
```
Why the `flat` in the name?

pwrobelse replied:
This function returns `std::vector<token>` - not `std::vector<dht::token_range>`.

`flat` was intended to indicate, that his is 1D and ranges are described by the adjacent elements. 1D array is used by `dht::describe_ownership()`.

That was my intention

avikivity replied:
Better to choose a new name, "flat" doesn't convey that.

pwrobelse replied:
Replaced with `get_sorted_tokens()`

denesb commented on locator/tablets.cc:
```diff
@@ -585,6 +607,23 @@ class tablet_effective_replication_map : public effective_replication_map {
     }
 
+    // FIXME: return a future object.
+    virtual dht::token_range_vector get_ranges(inet_address ep) const override {
```
Where is this used?

pwrobelse replied:
In `storage_service` class there is the following function. Previously, it was specific to `vnode_effective_replication_map` class. Now it uses `effective_replication_map`.

```
dht::token_range_vector         
storage_service::get_ranges_for_endpoint(const locator::effective_replication_map_ptr& erm, const gms::inet_address& ep) const {
    return erm->get_ranges(ep); 
}
```


It is used in a few places in `storage_service`.
One of such usages is in `storage_service::effective_ownership()`.

It looks as follows:

```
for (const auto& [dc, endpoints_map] : datacenter_endpoints) {
    for (auto endpoint : endpoints_map) {
        // calculate the ownership with replication and add the endpoint to the final ownership map
        try {
            float ownership = 0.0f;
            auto ranges = ss.get_ranges_for_endpoint(erm, endpoint);
            for (auto& r : ranges) {
                // get_ranges_for_endpoint will unwrap the first range.
                // With t0 t1 t2 t3, the first range (t3,t0] will be split
                // as (min,t0] and (t3,max]. Skippping the range (t3,max]
                // we will get the correct ownership number as if the first
                // range were not split.
                if (!r.end()) {
                    continue;
                }
                auto end_token = r.end()->value();
                auto loc = token_ownership.find(end_token);
                if (loc != token_ownership.end()) {
                    ownership += loc->second;
                }
            }
            final_ownership[endpoint] = ownership;
        }  catch (replica::no_such_keyspace&) {
            // In case ss.get_ranges_for_endpoint(keyspace_name, endpoint) is not found, just mark it as zero and continue
            final_ownership[endpoint] = 0;
        }
    }
}
```


I added `get_ranges(endpoint)` to `effective_replication_map` to force the derived classes to provide such functionality. Because of that I was able to reuse the logic in `storage_service::effective_ownership()` also for tablets.

denesb commented on api/api-doc/storage_service.json:
```diff
@@ -433,6 +433,14 @@
                      "allowMultiple":false,
                      "type":"string",
                      "paramType":"path"
```
There is too much going on in this commit, this should be split into like 4 different commits:
* add get_flat_token_ranges()
* make get_ranges() virtual
* change storage_service::effective_ownership()
* update API endpoint (maybe has to be folded into previous one if splitting is difficult)

pwrobelse replied:
Done - I split it into three patches:
- locator/tablets: add tablet_map::get_flat_token_ranges()
- locator/effective_replication_map: make 'get_ranges(inet_address ep)' virtual
- storage_service/ownership/{keyspace}: handle requests when tablets are enabled

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7014/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7014/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8193/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7014/)
:x: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/155/)
#### Build Details:
- Duration: 2 hr 10 min
- Builder: spider8.cloudius-systems.com

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7019/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7019/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/159/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7019/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8197/)
#### Build Details:
- Duration: 2 hr 15 min
- Builder: spider8.cloudius-systems.com

pwrobelse commented:
To ensure that I did not break anything, I run the following tests locally.

 Run in `tests/nodetool`:
  - 1: `pytest --nodetool=scylla .`
  - 2: `pytest --nodetool=cassandra .`
  
 Run via dtest:
  - 3: `nodetool_test.py`
  - 4: `nodetool_additional_test.py`
 
 They seem to work correctly.
 
 Note: in case of (2) I saw three issues in `test_compactionhistory.py` related to time mismatch in compared values (2 hours shift). It does not seem to be related to `storage_service/ownership`.
 

Note: in case of (4), when I run the tests from `scylla-dtest:next`, then I saw two failures on my local PC.
>  FAILED nodetool_additional_test.py::TestNodetool::test_nodetool_refresh_with_wrong_upload_modes
>  FAILED nodetool_additional_test.py::TestNodetool::test_nodetool_refresh_with_data_perms

denesb commented:
> To ensure that I did not break anything, I run the following tests locally.
> 
> Run in `tests/nodetool`:
> 
>     * 1: `pytest --nodetool=scylla .`
> 
>     * 2: `pytest --nodetool=cassandra .`
> 

These tests completely bypass ScyllaDB's REST API, so they are completely irrelevant for checking the fallout from a change in said REST API.

> 
> Run via dtest:
> 
>     * 3: `nodetool_test.py`
> 
>     * 4: `nodetool_additional_test.py`

These are indeed relevant.

> 
> They seem to work correctly.
> 
> Note: in case of (2) I saw three issues in `test_compactionhistory.py` related to time mismatch in compared values (2 hours shift). It does not seem to be related to `storage_service/ownership`.

If I recall correctly, this has something to do with timezone set in C locale.

> Note: in case of (4), when I run the tests from `scylla-dtest:next`, then I saw two failures on my local PC.
> 
> > FAILED nodetool_additional_test.py::TestNodetool::test_nodetool_refresh_with_wrong_upload_modes
> > FAILED nodetool_additional_test.py::TestNodetool::test_nodetool_refresh_with_data_perms

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7054/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7054/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/192/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7054/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8232/)
#### Failed Tests (1/23543):
- [alternator.run.release.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7054/testReport/junit/%28root%29/non-boost%20tests/alternator_run_release_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+alternator.run.release.2)
#### Build Details:
- Duration: 2 hr 24 min
- Builder: spider7.cloudius-systems.com

avikivity commented:
Please file/update issues for flaky tests.

pwrobelse commented:
> Please file/update issues for flaky tests.

Done:
 - https://github.com/scylladb/scylladb/issues/17564

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7065/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7065/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7065/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/203/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8243/)
#### Build Details:
- Duration: 2 hr 43 min
- Builder: spider4.cloudius-systems.com

pwrobelse commented:
> To ensure that I did not break anything, I run the following tests locally.
> 
> Run in `tests/nodetool`:
> 
>     * 1: `pytest --nodetool=scylla .`
> 
>     * 2: `pytest --nodetool=cassandra .`
> 
> 
> Run via dtest:
> 
>     * 3: `nodetool_test.py`
> 
>     * 4: `nodetool_additional_test.py`
> 
> 
> They seem to work correctly.
> 
> Note: in case of (2) I saw three issues in `test_compactionhistory.py` related to time mismatch in compared values (2 hours shift). It does not seem to be related to `storage_service/ownership`.
> 
> Note: in case of (4), when I run the tests from `scylla-dtest:next`, then I saw two failures on my local PC.
> 
> > FAILED nodetool_additional_test.py::TestNodetool::test_nodetool_refresh_with_wrong_upload_modes
> > FAILED nodetool_additional_test.py::TestNodetool::test_nodetool_refresh_with_data_perms


Regarding the tests run via dtest - the previous results were obtained without `--tablets` flag. I run the tests again with the mentioned flag.

nodetool_test.py passed without any troubles.
```
./scripts/run_test.sh --scylla-version='unstable/master:latest' nodetool_test.py --tablets
```

In case of nodetool_additional_test.py, the following three TCs failed: **test_concurrent_rebuild**, **test_nodetool_refresh_with_data_perms** and **test_nodetool_refresh_with_wrong_upload_modes**.
```
./scripts/run_test.sh --scylla-version='unstable/master:latest' nodetool_additional_test.py --tablets
```

The reasons of failure: 

- test_concurrent_rebuild - usage of `/storage_service/describe_ring` ([issue#16846](https://github.com/scylladb/scylladb/issues/16846)).

denesb commented:
@pwrobelse when pushing a new version of a PR, please also add a comment with a short changelog, describing what has changed in the new version, compared with the previous state of the code. This helps reviewers focus on the changed areas.

I now have to re-review this entire PR, if there was a changelog, I could focus on just the 1-2 commits that actually changed.

denesb reviewed (approved):
The code LGTM.

One more bureaucratic comment: you don't need to add Refs/Fixes to every commit in a series. It is better to just add any markers like that just to the cover letter. The exception is when a PR has a single commit, then indeed all these have to go to the commitlog of that single commit.

bhalevy commented:
@denesb is the PR ready to be merged?
If so, please do merge it.

avikivity commented on test/pylib/rest_client.py:
```diff
@@ -162,6 +162,17 @@ async def get_host_id_map(self, dst_server_ip: IPAddress) -> list[HostID]:
+        elif table is None:
+            api_path = f"/storage_service/ownership/{keyspace}"
+        else:
+            api_path = f"/storage_service/ownership/{keyspace}?cf={table}"
```
This breaks for keyspace is None and table is not None.

Also, it's excessive splitting, adding a simple unused function.

bhalevy commented on api/storage_service.cc:
```diff
@@ -1359,16 +1370,30 @@ void set_storage_service(http_context& ctx, routes& r, sharded<service::storage_
     ss::get_effective_ownership.set(r, [&ctx, &ss] (std::unique_ptr<http::request> req) {
-        auto keyspace_name = req->param["keyspace"] == "null" ? "" : validate_keyspace(ctx, req->param);
-        return ss.local().effective_ownership(keyspace_name).then([] (auto&& ownership) {
+        auto keyspace = req->param["keyspace"] == "null" ? "" : validate_keyspace(ctx, req->param);
```
Why is `keyspace_name` changed to `keyspace`?
Better reduce the churn if this is unnecessary.

pwrobelse replied:
Fixed :+1:

denesb commented:
The PR needs a rebase, there are conflicts.

kbr-scylla commented:
Dequeued, breaks build: https://github.com/scylladb/scylladb/issues/17699

kbr-scylla commented:
Please remember to run new tests 100 times, preferably in debug mode
(cc @scylladb/scylla-maint )

pwrobelse commented:
Changes between patches:
1. Aligned the PR with the latest master.
2. Corrected the names of variables in ss::get_effective_ownership() according to Benny's comments.
3. Adjusted the tests to use new fixtures instead of using `@pytest.mark.xfail`.
4. Removed the test of ownership that assumed equal number of tablets on each node - that is not guaranteed.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7230/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7230/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8400/)
:x: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/362/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7230/)
#### Build Details:
- Duration: 2 hr 1 min
- Builder: spider2.cloudius-systems.com

pwrobelse commented:
> #### 🔴 CI State: FAILURE
> 
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7230/) ✅ - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7230/) ✅ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8400/) ❌ - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/362/) ✅ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7230/)
> #### Build Details:
> 
>     * Duration: 2 hr 1 min
> 
>     * Builder: spider2.cloudius-systems.com

Created a new issue for the test that failed:
- https://github.com/scylladb/scylladb/issues/17704

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7271/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7271/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8431/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/397/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7271/)
#### Build Details:
- Duration: 2 hr 4 min
- Builder: spider1.cloudius-systems.com
